locator/abstract_replication_strategy.hh

```diff
@@ -98,6 +98,12 @@ public:
     virtual void validate_options() const = 0;
     virtual std::optional<std::set<sstring>> recognized_options() const = 0;
     virtual size_t get_replication_factor() const = 0;
+    // Decide if the replication strategy allow removing the node being
+    // replaced from the natural endpoints when a node is being replaced in the
+    // cluster. LocalStrategy is the not allowed to do so because it always
+    // returns the node itself as the natural_endpoints and the node will not
+    // appear in the pending_endpoints.
```

Comment:
I still think that having `get_natural_endpoints_without_pending()` at this level would be better, because you wouldn't have to understand what storage_proxy does in order to understand what this means.

Reply:
It is worse in my view because the abstract_replication_strategy.hh needs to understand pending nodes which only token_metadata knows.

Reply:
I think the concept of pending nodes belongs to replication strategy. We currently get it from  token_metadata, but it doesn't belong there. Notice how `get_pending_ranges()` takes a keyspace name, because each replication strategy may have a different idea about what pending ranges should be. `token_metadata_impl::calculate_pending_ranges_for_bootstrap()` takes an `abstract_replication_strategy&`. Eventually we should clean this up.

Reply:
I agree with Tomke here and this is what I proposed on the first version of the patchset. The function does not belong in the storage_proxy since it is not the one who is doing all the vnode calculations.

Reply:
> I think the concept of pending nodes belongs to replication strategy. We currently get it from token_metadata, but it doesn't belong there. Notice how `get_pending_ranges()` takes a keyspace name, because each replication strategy may have a different idea about what pending ranges should be. `token_metadata_impl::calculate_pending_ranges_for_bootstrap()` takes an `abstract_replication_strategy&`. Eventually we should clean this up.

I think you are mixing up pending node and pending ranges. Pending node is a per cluster.  Pending ranges are per keyspace that's why get_pending_ranges takes keyspace.

I think we can move the helper out of storage_proxy.

Reply:
> > I think the concept of pending nodes belongs to replication strategy. We currently get it from token_metadata, but it doesn't belong there. Notice how `get_pending_ranges()` takes a keyspace name, because each replication strategy may have a different idea about what pending ranges should be. `token_metadata_impl::calculate_pending_ranges_for_bootstrap()` takes an `abstract_replication_strategy&`. Eventually we should clean this up.
> 
> I think you are mixing up pending node and pending ranges. Pending node is a per cluster. Pending ranges are per keyspace that's why get_pending_ranges takes keyspace.

Pending ranges contain pending nodes. In this context, "pending node" is the node which is added to the replica set for a given token range. We can have no pending node in the ring (no physical node added), but can have pending nodes in replication strategy, e.g. when we decommission a node. This is the meaning of "pending node" (node = endpoint) which `token_metadata::pending_endpoints_for()` uses.

Pending ranges are actually per replication strategy instance. We currently have one replication strategy per keyspace, but when we introduce tablet partitioning, each table will have its own replication strategy and it will be per table.

Reply:
If you are talking about future it is a different story, we can change anything in the future. At the moment, the pending range is per keyspace. Take a look at _pending_ranges, _pending_ranges_interval_map.

Reply:
> Pending ranges contain pending nodes. In this context, "pending node" is the node which is added to the replica set for a given token range. We can have no pending node in the ring (no physical node added), but can have pending nodes in replication strategy, e.g. when we decommission a node. This is the meaning of "pending node" (node = endpoint) which token_metadata::pending_endpoints_for() uses.

Of course pending ranges have a node associated with the ranges, otherwise the ranges are meaningless. Then you should have said "pending ranges" not "pending node" belong to replication strategy.  However, the replication strategy does not need to know the pending ranges because, we only need replication strategy to tell what ranges belong to what node given a token ring, we can calculate the pending ranges. The given token ring is a parameter to replication strategy. If we move pending ranges to replication strategy, replication strategy needs to store a temporary token ring which is unnecessary and less flexible.

Reply:
> > Pending ranges contain pending nodes. In this context, "pending node" is the node which is added to the replica set for a given token range. We can have no pending node in the ring (no physical node added), but can have pending nodes in replication strategy, e.g. when we decommission a node. This is the meaning of "pending node" (node = endpoint) which token_metadata::pending_endpoints_for() uses.
> 
> Of course pending ranges have a node associated with the ranges, otherwise the ranges are meaningless. Then you should have said "pending ranges" not "pending node" belong to replication strategy. 

I think I can say that the concept of a pending node belongs to replication strategy at the API level. It doesn't matter how you internally organize data. You ask replication strategy for pending nodes for a given key, so the user needs to understand what a pending node is, so this concept belongs to the idea of replication strategy.

> However, the replication strategy does not need to know the pending ranges because, we only need replication strategy to tell what ranges belong to what node given a token ring, we can calculate the pending ranges. The given token ring is a parameter to replication strategy. If we move pending ranges to replication strategy, replication strategy needs to store a temporary token ring which is unnecessary and less flexible.

The fact that replication strategy relies on token ring for replication is a detail of the strategy. That's why I think pending ranges should belong to the replication strategy and not token ring. 

There could be strategies which don't rely on the ring, like tablet partitioning, and they would obtain pending nodes from somewhere else. 

The user of the strategy should not be concerned with where the strategy gets its replication information from, he should just be able to ask it for old and new replica sets for a given key.

So it's more flexible if we don't store pending ranges in token metadata but in the strategy.